### DISCUSSION TITLE
select with explicit type casts

### DISCUSSION THREAD

**Discussion by bluebill1049** [supabase/supabase#3582] [Questions] [answered]
Is the following query possible to achieve with [select](https://supabase.io/docs/reference/javascript/select) API?

```sql
select title from table_a
where 
data::text like '%something alike%'
```

the `data` column's type is `jsonb`, the query works fine by itself and I can potentially turn it into a `function` just wondering type cast is possible with `select` alone.

**Comment by steve-chavez** [comment, marked as answer]
>  wondering type cast is possible with select alone

That's not possible because casting the filter column would invalidate any indexes you set up - allowing your API clients to force inefficient queries.

> I can potentially turn it into a function 

You could save a full RPC in this case, with a [computed column](https://postgrest.org/en/v8.0/api.html#computed-columns)

```sql
create function data_text(table_a) returns text as $$
  select $1.data::text;
$$ language sql;

-- you can also index the computed column
```
Then call it like
```js
const { data, error } = await supabase
  .from('table_a')
  .select()
  .like('data_text', '..'):
```

**Reply by bluebill1049** [reply]
Thanks @steve-chavez for the detailed reply and yes I have seen your solution in another discussion as well. Wondering what's the tradeoff with this approach, let's say if the JSON gets relatively large, would that impact perf on each upsert?

**Reply by steve-chavez** [reply]
> Wondering what's the tradeoff with this approach, let's say if the JSON gets relatively large, would that impact perf on each upsert?

Hm, you mean the tradeoff of using an index on a jsonb column vs a text column?

I'm not sure, but in general adding indexes will incur on a performance penalty for inserts because the index has to be updated.

**Reply by bluebill1049** [reply]
Thanks Steve for the reply and help. 🙏
